src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala

```diff
@@ -109,28 +112,28 @@ class PetEndpoints[F[_]: Effect] extends Http4sDsl[F] {
 
   private def findPetsByStatusEndpoint(petService: PetService[F]): HttpService[F] =
     HttpService[F] {
-      case GET -> Root / "pets" / "findByStatus" :? StatusMatcher(Valid(Nil)) =>
-        // User did not specify any statuses
-        BadRequest("status parameter not specified")
-
       case GET -> Root / "pets" / "findByStatus" :? StatusMatcher(Valid(statuses)) =>
         // We have a list of valid statuses, find them and return
-        for {
-          retrieved <- petService.findByStatus(NonEmptyList.fromListUnsafe(statuses))
+        (for {
+          statusesNel <- AE.fromEither(statuses.toNel.toRight(EmptyStatuses))
+          retrieved <- petService.findByStatus(statusesNel)
           resp <- Ok(retrieved.asJson)
-        } yield resp
+        } yield resp).recoverWith {
+        case EmptyStatuses => BadRequest("tag parameter not specified")
+      }
     }
 
+
   private def findPetsByTagEndpoint(petService: PetService[F]): HttpService[F] =
     HttpService[F] {
-      case GET -> Root / "pets" / "findByTags" :? TagMatcher(Valid(Nil)) =>
-        BadRequest("tag parameter not specified")
-
       case GET -> Root / "pets" / "findByTags" :? TagMatcher(Valid(tags)) =>
-        for {
-          retrieved <- petService.findByTag(NonEmptyList.fromListUnsafe(tags))
-          resp <- Ok(retrieved.asJson)
-        } yield resp
+        (for {
+          tagsNel   <- AE.fromEither(tags.toNel.toRight(EmptyTags))
```

Comment:
We do the same thing here twice.  

Honestly, I think this should come out of the box with http4s.  We need a required, mult query param decoder.

Perhaps we can implement our own "RequiredMultiQueryParamDecoderMatcher" and contribute it back to http4s?

Reply:
That's a great idea! I'll look into that :)